OSPR-535 Partial Credit

Colin-Fredericks opened the pull request:
**Background:** This project adds partial credit grading schemes to four different problem types (MC, Checkbox, Dropdown, Numerical), allows Custom Python-Graded problems to return "partially correct" responses, and makes it easier for partial credit to be built into other problem types in the future. This change is 100% backward-compatible; no old problems will need to be changed.

**EdX Contacts:** @sarina , @explorerleslie , @pmitros , and @marcotuts have been consulted on this particular project.

**Outside Discussion:** http://help.edge.edx.org/discussions/suggestions/431-request-for-partial-credit

**JIRA ticket (proposal):** https://openedx.atlassian.net/browse/OSPR-535
**JIRA ticket (review):** https://openedx.atlassian.net/browse/OSPR-680

**Affected components:** Capa problems, including their templates, user feedback, and grading code. The visual changes are very subtle - the only indication to a sighted user of partial credit is the number of points shown.

**LMS updates:** Described in detail with screenshots in the following google doc:
https://docs.google.com/document/d/12CZwqG5WdQreFoZMiIxezQMh5KDtjHy7xEWKwJ71F9w/edit?usp=sharing

**Setting this up on your own devstack:** Here's a [course tarball](https://github.com/Colin-Fredericks/edx-partial-credit/raw/master/PartialCreditCourse.tar.gz) you can use via Studio's "Import" function. No setting changes or outside servers are needed. Sample XML is also provided in the document above, which can then be placed inside `<problem></problem>` tags in a Blank Advanced Problem.

A .tar.gz file of a Studio course with example problems is here: _TODO_

**Author Concerns:** I'm still worried that students won't realize they've received partial credit if they're still seeing the green checkmark, but I'm ok leaving this decision (and future changes to it) in the hands of edX's UX team.

Revised after first pull discussion here: https://github.com/edx/edx-platform/pull/8574

edx-webhook commented:
Thanks for the pull request, @Colin-Fredericks! I've created [OSPR-680](https://openedx.atlassian.net/browse/OSPR-680) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've submitted a [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf) or indicated your institutional affiliation and added yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

Colin-Fredericks commented:
@sarina, @pmitros New version of old PR.

I'm a little nervous about the amount of change to the scss file since the last one I put together.

sarina commented:
I think Marco made those changes so if you have questions I recommend
reaching out to him.
On Jun 30, 2015 12:14 PM, "Colin-Fredericks" <notifications@github.com>
wrote:

> @sarina <https://github.com/sarina>, @pmitros <https://github.com/pmitros>
> New version of old PR.
>
> I'm a little nervous about the amount of change to the scss file since the
> last one I put together.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8724#issuecomment-117241855>.
>

Colin-Fredericks commented:
Thanks for the suggestion. @marcotuts , would you mind looking through the changes I made in common/lib/xmodule/xmodule/css/capa/display.scss and making sure they fit properly with the new style for all this? I tried to copy existing styles and only add what was necessary for partially-correct support.

Colin-Fredericks commented:
Ok, good news - I fixed a bunch of typos in my scss file, AND updated my devstack so that all the new recent stuff is pulled in (like the new appearance for checkbox-style problems).

Bad news - A lot of that recent stuff isn't working well in my devstack, even with the fixed scss. (Boy I wish I had started a month earlier so they had to fix their code to work with mine mid-development...) And the *really* bizarre part is that in many problem types, *partially correct problems are displaying correctly, and correct or incorrect problems are not*, and that's super-weird.

More on this as I figure out what's going wrong.

Colin-Fredericks commented:
Got it! It was an issue with the templates, and an extra block in the scss that didn't need to be there any more.

Colin-Fredericks commented:
Successful rebase! Thanks @sarina for the tutorial.

Unfortunately, now the build is just timing out here. (I'm having flashbacks to compiling C++ and getting 2374 errors because I missed one semicolon...) Luckily, it doesn't time out on my own devstack. I have one error (which will be fixed soon) with all-wrong checkbox problems showing up as correct, and eight failed tests that have to do with the new hints functionality on text input problems.

I'm not sure what to do about the remaining errors. They seem to be from code that I didn't alter and that shouldn't depend on the (very minor) changes I made to the textline.html template.

sarina commented:
The build times out because our Jenkins server falls under incredible load
between 9 am - 6 pm EDT. If you're continually getting build timeouts I
suggest you try pushing a new commit outside of this time range. pushing
new commits OR rebasing your branch will trigger new builds.

On Mon, Jul 6, 2015 at 12:30 PM, Colin-Fredericks <notifications@github.com>
wrote:

> Successful rebase! Thanks @sarina <https://github.com/sarina> for the
> tutorial.
>
> Unfortunately, now the build is just timing out here. (I'm having
> flashbacks to compiling C++ and getting 2374 errors because I missed one
> semicolon...) Luckily, it doesn't time out on my own devstack. I have one
> error (which will be fixed soon) with all-wrong checkbox problems showing
> up as correct, and eight failed tests that have to do with the new hints
> functionality on text input problems.
>
> I'm not sure what to do about the remaining errors. They seem to be from
> code that I didn't alter and that shouldn't depend on the (very minor)
> changes I made to the textline.html template.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8724#issuecomment-118917869>.
>

Colin-Fredericks commented:
Good to know. I'll give it a shot off-hours.

sarina commented:
Your build also may fail because of quality. Click the "Diff Quality"
report on the left side of your build page, and run `paver run_quality`
locally to get a local report. The local report may not match the Jenkins
report; if that happens, run `paver run_pylint` prior to running the `paver
run_quality` command. See
https://github.com/edx/edx-platform/blob/master/docs/en_us/internal/testing.rst#python-code-style-quality

On Mon, Jul 6, 2015 at 1:18 PM, Sarina Canelake <sarina@edx.org> wrote:

> The build times out because our Jenkins server falls under incredible load
> between 9 am - 6 pm EDT. If you're continually getting build timeouts I
> suggest you try pushing a new commit outside of this time range. pushing
> new commits OR rebasing your branch will trigger new builds.
>
> On Mon, Jul 6, 2015 at 12:30 PM, Colin-Fredericks <
> notifications@github.com> wrote:
>
>> Successful rebase! Thanks @sarina <https://github.com/sarina> for the
>> tutorial.
>>
>> Unfortunately, now the build is just timing out here. (I'm having
>> flashbacks to compiling C++ and getting 2374 errors because I missed one
>> semicolon...) Luckily, it doesn't time out on my own devstack. I have one
>> error (which will be fixed soon) with all-wrong checkbox problems showing
>> up as correct, and eight failed tests that have to do with the new hints
>> functionality on text input problems.
>>
>> I'm not sure what to do about the remaining errors. They seem to be from
>> code that I didn't alter and that shouldn't depend on the (very minor)
>> changes I made to the textline.html template.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/edx/edx-platform/pull/8724#issuecomment-118917869>.
>>
>
>

sarina commented:
@Colin-Fredericks looking at your build, https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/11293/, you are failing the diff-quality report as well as all the capa failures are the result of your changes. I think the JavaScript test failure is unrelated so you may ignore it.

Colin-Fredericks commented:
The pylint violation is easily fixed since I don't think I wrote that block of code in the first place. Must be an artifact from a manual merge. Poking around on my devstack indicates that might be what's causing the hint-testing errors as well. Will push.

Colin-Fredericks commented:
Success!!!

Will rebase, now that I have that workflow going properly.

Now on to writing tests...

Colin-Fredericks commented on common/lib/capa/capa/tests/response_xml_factory.py:
```diff
@@ -193,6 +230,12 @@ def create_response_element(self, **kwargs):
+                responseparam_element.set('partial_range', str(partial_range))
+
+        if partial_answers is not None and 'list' in credit_type:
+            responseparam_element = etree.SubElement(response_element, 'responseparam')
```
Line 237 here is throwing a pylint error. Line 230 is not. I have no idea why, since I literally copy-pasted this one from that one. Any ideas?

The error is `237: E1101: (no-member), NumericalResponseXMLFactory.create_response_element: Module 'safe_lxml.etree' has no 'SubElement' member`

sarina replied:
@Colin-Fredericks I suspect this pylint error is actually present elsewhere in the file; try running simply `pylint path/to/file` locally to see all the pylint errors in the whole file. It would be great if you cleaned up identical violations in otherwise untouched lines (straightening up as you go) ,but it's not required.

That said this is a weird pylint error. @benpatterson do you know anything about this?

benpatterson replied:
@sarina @Colin-Fredericks Indeed that file has these violations already. See report on your last build [here](https://build.testeng.edx.org/job/edx-platform-test-subset/131503/artifact/reports/common/pylint.report/*view*/). The intention of using diff-quality is so that, at the least, we don't continue a bad pattern. So, you ought to fix it on the new lines.

That said, this particular error can give false positives, as pointed out [here](http://pylint-messages.wikidot.com/messages:e1101). I'm only glancing at this code, but if this is a dynamically created member, then perhaps this applies to you.

benpatterson replied:
I should also mention...I admit I try to use it sparingly, but you can disable pylint violations inline. Very similar example in [this thread](https://groups.google.com/forum/#!topic/edx-code/qNk29Z5QnfA).

Colin-Fredericks replied:
@benpatterson Thanks for the info. Since that error shows up 41 times in the link you gave me for the last build, I'm going to propose that maybe this error is full of crap. I'm inclined to disable inline with a comment to explain why.

Colin-Fredericks commented:
94% test coverage! Almost there. Need a little more for Numerical and several more for Custom.

The one line missed in correctmap.py will be fixed next pull, which will probably be a rebase.

marcotuts commented:
@Colin-Fredericks  Thanks for the ping, and thank you for making the changes given the recent capa styling changes + the hinting changes! The styles at a glance (i.e. without reviewing a sandbox) looked good to me. I see that you did still have some image based feedback elements, which we've tried to move away from toward font-awesome, but since the hinting work does use images I think you did the right thing to use the approach currently in use by that part of capa. I would love to review eventually on a sandbox whenever that is possible but for now I didn't have any suggested changes.

Thanks!

Colin-Fredericks commented:
Thanks, @marcotuts. I appreciate you looking it over.

On my local devstack I ended up switching the partially-correct css styles to use a different FontAwesome icon so that I could see more quickly whether something was being marked with partial credit. (I used https://fortawesome.github.io/Font-Awesome/icon/star-half-o/ in orange.) The version here uses the green check for both partially-correct and fully-correct.

Colin-Fredericks commented:
Remaining items:
Three untested lines, all in CustomResponse
One pylint issue (the one mentioned above)

...and I think that might be it. Am I forgetting anything that I will need to do before this goes to the review/documentation phase?

sarina commented:
Please be sure your PR has a good cover letter including manual testing
instructions. See
http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/process/cover-letter.html

On Wed, Jul 8, 2015 at 3:36 PM, Colin-Fredericks <notifications@github.com>
wrote:

> Remaining items:
> Three untested lines, all in CustomResponse
> One pylint issue (the one mentioned above)
>
> ...and I think that might be it. Am I forgetting anything that I will need
> to do before this goes to the review/documentation phase?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8724#issuecomment-119708070>.
>

Colin-Fredericks commented:
Victory!
https://www.youtube.com/watch?v=pCLIg6WT7zM

Colin-Fredericks commented:
So - all tests passing, 100% test coverage, 100% code quality, cover letter written, XML documented, screen shots of UI updated. What's the next step? How can I help?

pmitros commented:
:+1: I'm good with it. There are things I would have preferred to see a little differently (namely, the router in LoncapaResponse, rather than inside of each individual ResponseType, and less code replication), but from my point of view, this is clearly better than before, so I'm okay leaving that for a future PR.

One request (but not requirement): If you could document future code path for future developers, that would help this evolve more cleanly. A comment to the effect of:

> Planned code evolution path: We would like to move the routing code for partial credit into LoncapaResponse. This will allow us to implement things like `every_decision_counts` only once, and use common partial credit graders, and remove a lot of code replication. 

Would help a lot.

Colin-Fredericks commented:
@pmitros Thanks! I'll put that into the Google Drive document, since there are a bunch of "v2" items there as well.

sarina commented:
@Colin-Fredericks this will need to go through code review which may take a little while to schedule, because a lot of people will need to look at it (T&L, accessibility, documentation, mobile, and UX). I will start coordinating it.

Also the PR needs rebasing because there is some conflict with master.

Colin-Fredericks commented:
Thanks for the info. I'm happy to come over to edX and show this to folks if that would help. I'm around a lot this week, but really swamped next week and away at a conference the week after.

Should I do the rebase right away, or should I wait until we're closer to pulling so that we don't have to do it multiple times?

sarina commented:
The rebasing is up to you. I often find that frequent rebasing is preferable, because the issues are small and it's easy to figure out when a change was made that my code conflicts with, and it's easy to inspect the small area that conflicted to ensure I solved the rebase properly. Rebasing after a few weeks can end up with a ton of conflicts that are annoying to have to solve all at once and more challenging to validate. That said, your call.

Colin-Fredericks commented:
Thanks. I'll try to keep it frequent, then.

sarina commented:
Also if you sign up for [Julython](http://www.julython.org/), you'll get a point for a commit with every rebase! :grin:

sarina commented:
@Colin-Fredericks Instead of this ("Sample XML is provided in the document above, which can then be placed inside <problem></problem> tags in a Blank Advanced Problem."), can you please create a new course on your instance with examples of your problem, then export the course using the Studio exporter. Then, upload that file into your pull request description so that people can quickly and easily test the new functionality.

sarina commented on common/lib/capa/capa/tests/test_responsetypes.py:
```diff
@@ -1106,6 +1245,96 @@ def test_checkbox_group_grade(self):
+        # 'bongo' is not a valid grading scheme.
+        problem = self.build_problem(choice_type='checkbox',
+                                     choices=[False, False, True, True],
+                                     credit_type='bongo')
```
Multiline breaks should be like this:
```python
problem = self.build_problem(
    choice_type='checkbox',
    choices=[False, False, True, True],
    credit_type='bongo'
)
```

See https://github.com/edx/edx-platform/wiki/Python-Guidelines#breaking-long-lines

sarina replied:
I won't comment on all instances but this is a cleanup that you should do throughout the ocde.

Colin-Fredericks replied:
Can do.

sarina commented:
@Colin-Fredericks code looks good at a very brief pass. Once I have a tarball of a course, I'll set up a sandbox and do a functional review, then get it in team backlog. Teams work on two week sprint cycles and do have other commitments but hopefully I'll be able to get you an estimate of when the review will be scheduled.

For an example of a really great tarball that showed off a course feature, check out the tarball provided in the description of #6786. Note that he provided examples of how hinting was used on all problem types. Obviously your feature isn't as extensive, but including both types of partial credit, and writeups within the problems of what you're demonstrating, will be really helpful. Think of it as an edX course to explain partial credit! This will help all the review teams out greatly, particularly documentation and accessibility (because you're showing off all the different ways the feature can be used).

sarina commented:
You're going to be asked as well to write a bok choy test for your work. Bok choy acceptance tests are expensive so the philosophy is to use these tests as a happy path once over test. Test definitions are in `common/test/acceptance/tests/lms/test_lms_problems.py` (again check out what was added in #6786). More information is at https://github.com/edx/edx-platform/blob/master/docs/en_us/internal/testing.rst#ui-acceptance-tests (information on running bok choy locally is later in this document).

Colin-Fredericks commented:
[Course tarball](https://github.com/Colin-Fredericks/edx-partial-credit/raw/master/PartialCreditCourse.tar.gz) added to the cover letter at the top of this page (and it's also that link). Luckily for me I basically already had this together.

Will get working on Bok Choy.

benpatterson commented:
@Colin-Fredericks

You may find the tutorial for bok-choy helpful. Documentation here: 

http://bok-choy.readthedocs.org/en/latest/

Colin-Fredericks commented:
@benpatterson Thanks for the link. I've got a test mocked up fairly well, and it's passing (after writing some tools that work with the new problem HTML templates). It creates a numerical problem, puts in a value that's partially correct, and checks to see whether the right status appears. 

Anything else I should be checking for this, or does that basically do it? The rest of the tests I looked at seemed similarly brief.

benpatterson commented on common/test/acceptance/pages/lms/problem.py:
```diff
@@ -68,6 +74,24 @@ def is_correct(self):
 
+    def simpleprob_is_correct(self):
+        """
+        Is there a "partially correct" status showing? Works with simple problem types.
```
I'm wondering if these simpleprob_is_* methods need to have updated docstrings. The descriptions are the same, but I'm guessing from the names they are different (for good reasons)... Could you update these to more clearly draw the distinctions between them?

Colin-Fredericks replied:
Whoops - yes, copypasta going on there. Will fix.

benpatterson commented on common/test/acceptance/tests/lms/test_lms_problems.py:
```diff
@@ -161,3 +161,35 @@ def test_demand_hint(self):
+    """
+    Makes sure that the partial credit is appearing properly.
+    """
+    def get_problem(self):
```
I don't see this method being called by the test_ method. Should this method be part of setUp? Or removed?

Colin-Fredericks replied:
The base class (ProblemsTest, at the top of this file) is the one that calls self.get_problem() (line 36 at the moment). If you check out ProblemExtendedHintTest and ProblemClarificationTest, they also don't explicitly call their get_problem() methods.

Does that help, or were you asking about something else?

benpatterson replied:
Ah, great. Thanks @Colin-Fredericks.  I wonder if that's missing the `@abstractmethod` decorator, but that can be covered separately...not a concern for this PR.

benpatterson commented:
@Colin-Fredericks overall I think the bok-choy test looks good. I had some questions in sorting it out. I'm also wondering if -- whether it's via bok-choy or unit tests -- you confirm when an answer is *not* partially-correct. I agree with @sarina it's much cheaper to handle that in the unit tests. I think I see it being tested in one of the problem types, but I just wanted to doublecheck that negative test is covered across the problem types.

But back to the bok-choy, it looks good overall. I think just pointing out the pathway I'm missing would help me out, and updating the docstrings should happen. Now that I have a chance to get a longer look at this PR, I'm excited to see it :)

sarina commented:
@marcotuts have you signed off on the UX on this or do you need to provide a review? Please let me know.

cptvitamin commented:
I have found no additional accessibility issues with this PR :+1: 
nice work on this.

Colin-Fredericks commented:
@benpatterson I just went through the unit tests to double-check, and they do test to make sure that correct, incorrect, and partially-correct answers are returned properly when partial credit is enabled for all the affected problem types.

benpatterson commented:
Thanks @Colin-Fredericks . This gets a :+1: from me as far as bok-choy goes.

Colin-Fredericks commented:
![allyourrebase](https://cloud.githubusercontent.com/assets/1929508/8834747/1e89b9bc-3084-11e5-96f3-08500db3117f.png)

marcotuts commented:
@sarina  - I just saw your message above, my apologies for missing the tag before I was out for a while. I'm going to spin up a sandbox for this work to review the styling directly as the second part of my review. Thanks @Colin-Fredericks, and apologies for the delay here!

marcotuts commented:
@Colin-Fredericks  - I ran into one issue while testing the checkboxes problem you included as example XML in your google drive document while testing this one a sandbox. It appears that partial-credit problems aren't having their data-tooltip value set, which breaks the tooltips intended to communicate the status of the problem when learners hover over the icons. 

Here is a direct link to that problem along with a screenshot of the issue: 

([Link to PRoblem](http://pr8724.m.sandbox.edx.org/courses/course-v1:edx+102+120/courseware/02f93ffe02cc48539ebeedf282b8f17e/006e02bf473244f4a5050090190dde20/))

Screenshot:
![image](https://cloud.githubusercontent.com/assets/2023680/9027850/d0ae6cd0-3932-11e5-8e61-691a3b7669dd.png)

marcotuts commented on common/lib/capa/capa/inputtypes.py:
```diff
@@ -85,6 +85,7 @@ def __init__(self, status, gettext_func=unicode):
         names = {
             'correct': _('correct'),
             'incorrect': _('incorrect'),
+            'partially-correct': _('partially correct'),
```
a few lines below this section is the "tooltips" section, which needs one more line similar to this one to build the "This is partially correct" status tooltip.

Colin-Fredericks replied:
No problem. Fixing right now.

Colin-Fredericks commented:
@marcotuts Should be all taken care of. I just tested it out on my devstack and the tooltips are working properly now.

marcotuts commented:
@Colin-Fredericks Thanks for the fixes! I'm not able to test this out locally for some reason unfortunately. I'm seeing an issue with even rendering these capa problems locally. Have you run across anything like this before? 
![image](https://cloud.githubusercontent.com/assets/2023680/9038705/610f6d74-39c4-11e5-835d-3d8c35484dc3.png)

Thanks!

Colin-Fredericks commented:
Weird. I haven't seen that before. @sarina , @explorerleslie , @cptvitamin , did you see this happening in the sandboxes that you used?

@marcotuts I'm trying to get my devstack updated to make sure I'm using the same files you are, but the update process turned from what should be a single command in a terminal into a really unbelievable mess, so it may be a little while.

Colin-Fredericks commented:
After [a fair amount of work](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/edx-code/ffXvbZvOS_I/gmxtXBcvBwAJ) updating my devstack today, I'm still not having any rendering issues with these files.

sarina commented:
Marco can you just try again? Maybe use the course tarball Colin provided?

On Mon, Aug 3, 2015 at 4:44 PM, Colin-Fredericks <notifications@github.com>
wrote:

> After a fair amount of work
> <https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/edx-code/ffXvbZvOS_I/gmxtXBcvBwAJ>
> updating my devstack today, I'm still not having any rendering issues with
> these files.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8724#issuecomment-127401735>.
>

marcotuts commented:
I updated the sandbox and it looks like it is working again, I was able to confirm the partial credit tooltip was there. 

![image](https://cloud.githubusercontent.com/assets/2023680/9068733/2d0a231c-3ab3-11e5-9525-c5efda6b149e.png)

This just needs a rebase and it should be ready to go. (@Colin-Fredericks let me know if you have any questions during the rebase. We recently merged in a PR to fix a capa styling issue on annotation problems so my guess is that may overlap with your work. Here is a link to that in case it is helpful (https://github.com/edx/edx-platform/pull/9168) 

:+1: :rocket: :100:

Colin-Fredericks commented:
Awesome! Thanks.

The rebase was really easy - the only conflict was that both of us added a test to the end of the same file.

dianakhuang commented on common/lib/capa/capa/responsetypes.py:
```diff
@@ -813,12 +826,22 @@ class ChoiceResponse(LoncapaResponse):
+        )
+
+        self.incorrect_choices = set([choice.get(
+            'name') for choice in incorrect_xml])
```
Nit: it would probably be easier to read this if it were formatted more like: 
```
set([
     choice.get('name') for choice in incorrect_xml
])
```

Colin-Fredericks replied:
This was adapted from line 835 above. Happy to change if it would help.

I feel like my next project should be fixing all the pep8/pylint violations, hard-to-read code, and one-off bizarro functions that I found while adapting the existing code.

dianakhuang commented on common/lib/capa/capa/responsetypes.py:
```diff
@@ -831,26 +854,146 @@ def assign_choice_names(self):
+            student_answer = [student_answer]
+
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but score feedback is broken when no boxes are marked.
```
This comment is a little confusing because it's not quite clear what the current behavior is. When a student doesn't mark any boxes, their answer will be marked as incorrect, right?

Colin-Fredericks replied:
Yes - when a student marks no boxes, their submission is graded as incorrect, even if all of the choices are marked as "False". This didn't seem like good behavior to me, so I tried to fix it, but (as it says) things break if there are no boxes marked.

dianakhuang replied:
Thanks! That's good information to know. It would be great if you you could put that information into your comment.

Colin-Fredericks replied:
Done. Rebasing now.

dianakhuang commented on common/lib/capa/capa/responsetypes.py:
```diff
@@ -831,26 +854,146 @@ def assign_choice_names(self):
+        Calculates partial credit on the Every Decision Counts scheme.
+        For each correctly selected or correctly blank choice, score 1 point.
+        Divide by total number of choices.
+        Arguments: all_choices, student_answer, student_non_answers
```
Could you document what `all_choices`, `student_answer`, and `student_non_answers` are supposed to be (types, etc). I wasn't sure what `student_non_answers` was supposed to be referring to until I read the code further down where we generate them.

Colin-Fredericks replied:
Line 940 has this; I'm happy to move it if you prefer. Where would you like it?

dianakhuang replied:
I would say in the docstrings of any method that expect them as arguments. It's okay if it just gets duplicated a whole bunch. I'd much rather have that redundancy in documentation.

Colin-Fredericks replied:
No problem. Will do.

dianakhuang commented on common/lib/capa/capa/responsetypes.py:
```diff
@@ -1082,16 +1252,80 @@ def late_transforms(self, problem):
+
+        # Only one type of credit at a time.
+        if len(self.credit_type) > 1:
+            raise LoncapaProblemError('Only one type of partial credit is allowed for Multiple Choice problems.')
```
Are we concerned about `len(self.credit_type)` being less than one at all?

Colin-Fredericks replied:
No. Does len() ever return negatives?

sarina replied:
No, but len() can return 0

On Fri, Aug 7, 2015 at 3:28 PM, Colin-Fredericks <notifications@github.com>
wrote:

> In common/lib/capa/capa/responsetypes.py
> <https://github.com/edx/edx-platform/pull/8724#discussion_r36552870>:
>
> > +            return self.grade_without_partial_credit(student_answers=student_answers)
> > +
> > +        # This below checks to see whether we're using an alternate grading scheme.
> > +        #  Set partial_credit="false" (or remove it) to require an exact answer for any credit.
> > +        #  Set partial_credit="points" to set specific point values for specific choices.
> > +
> > +        # Translators: 'partial_credit' and the items in the 'graders' object
> > +        # are attribute names or values and should not be translated.
> > +        graders = {
> > +            'points': self.grade_via_points,
> > +            'false': self.grade_without_partial_credit
> > +        }
> > +
> > +        # Only one type of credit at a time.
> > +        if len(self.credit_type) > 1:
> > +            raise LoncapaProblemError('Only one type of partial credit is allowed for Multiple Choice problems.')
>
> No. Does len() ever return negatives?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8724/files#r36552870>.
>

Colin-Fredericks replied:
Just on suspicion, I tested what happens if someone enters `partial_credit=""` (empty string) or `partial_credit=" "` (whitespace). Luckily, it throws the same error then as it does if you specify a bogus credit type, so we probably don't have to worry about that length being zero. It'll only come out as 1 or more than 1.

dianakhuang commented on common/lib/capa/capa/correctmap.py:
```diff
@@ -101,10 +101,23 @@ def set_dict(self, correct_map):
+        Returns true if the problem is partially correct.
+        """
+        if answer_id in self.cmap:
+            return self.cmap[answer_id]['correctness'] in ['partially-correct']
```
Why can't this be `self.cmap[answer_id]['correctness'] == 'partially-correct'`?

Colin-Fredericks replied:
Could do that. I just copied line 109 and took out 'correct'.

dianakhuang commented:
Minus those few comments, this looks good to me. Kudos on wrestling with capa! I know it can be a beast to work with. Let me know when you want another review pass.

Colin-Fredericks commented:
Thanks @dianakhuang ! I'm rebasing now; the style comments you made are taken care of.

Since the `len(self.credit_type)` returning zero is caught by the next error (a valid credit type will never be found in a list of length zero) I think we're ok without explicitly checking that.

dianakhuang commented:
jenkins run bokchoy

sarina commented:
jenkins run bokchoy

sarina commented:
jenkins run bokchoy

sarina commented:
@dianakhuang Tests all pass; in your hands now.

Colin-Fredericks commented:
Thank you everyone for your help with this. I really appreciate it.

sarina commented:
Icon change ref: #9529
